apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AbstractAsyncHttpClientInstrumentation.java

```diff
@@ -179,6 +179,14 @@ public AsyncHandlerOnCompletedInstrumentation() {
 
         @Advice.OnMethodEnter(suppress = Throwable.class)
         private static void onMethodEnter(@Advice.This AsyncHandler<?> asyncHandler) {
```

Comment:
Making sure that wrapped `AsyncHandler`s work as expected (activate and deactivate only in outer-most `AsyncHandler`
```suggestion
        private static void onMethodEnter(@Advice.This AsyncHandler<?> asyncHandler, @Advice.Local("span") Span span) {
```

Reply:
Bit unsure on what you mean with activate and deactive in the outer-most AsyncHandler?

Reply:
Following the decorator pattern, calling `AsyncHandler#onComplete` may result in the handler calling another, wrapped/nested `AsyncHandler#onComplete`. We want to make sure that only the first `AsyncHandler` activates and deactivates the span.

In your first draft, the first `AsyncHandler` would activate the span but the last `AsyncHandler` in the chain would deactivate it. This would result in the span not being active in the first `AsyncHandler` after it called `onComplete` on its delegate handler.


```
    private final AsyncHandler<Response> delegate;

    @Override
    public Response onCompleted(Response response) {
        try {
            return delegate.onCompleted(response);
        } finally {
             tracer.getActive() // returns null
        }
    }
```

By propagating the active span via a `@Advice.Local`, we have fixed that problem.